bindings/matrix-sdk-ffi/src/client.rs

```diff
@@ -1101,13 +1102,38 @@ fn gen_transaction_id() -> String {
 /// is dropped, the file will be removed from the disk.
 #[derive(uniffi::Object)]
 pub struct MediaFileHandle {
-    inner: SdkMediaFileHandle,
+    inner: RwLock<Option<SdkMediaFileHandle>>,
+}
+
+impl MediaFileHandle {
+    fn new(handle: SdkMediaFileHandle) -> Self {
+        Self { inner: RwLock::new(Some(handle)) }
+    }
 }
 
 #[uniffi::export]
 impl MediaFileHandle {
     /// Get the media file's path.
     pub fn path(&self) -> String {
-        self.inner.path().to_str().unwrap().to_owned()
+        self.inner
+            .read()
+            .unwrap()
+            .as_ref()
+            .expect("MediaFileHandle must not be used after calling persist")
+            .path()
+            .to_str()
+            .unwrap()
+            .to_owned()
+    }
+
+    pub fn persist(&self, path: String) -> bool {
+        let mut guard = self.inner.write().unwrap();
+        match guard.take().expect("MediaFileHandle was already persisted").persist(path.as_ref()) {
```

Comment:
The `expect` here and above are a bit scary, can we:

- either have `persist` take ownership of `self`?
- or have both methods be fallible and return a `Result<..., ClientError>`?

Reply:
Re. taking ownership: No, UniFFI only supports `&self` and `Arc<Self>` (plus `&Arc<Self>`, I guess?) methods.

Re. using `Result`, I personally don't like it very much, it's a programming error just like indexing with an out-of-bounds index / range, which doesn't require acknowledging the error case in almost all programming languages, including Rust. UniFFI will catch the panic and convert it to an exception either way, this isn't undefined behavior or anything like that.

Reply:
In terms of personal preferences: when I debug rageshakes, I pay much more attention to panics (by explicitly starting to look for `panic`), since they usually indicate something gone very wrong, like breaking an internal assumption. Here it's a matter of breaking external assumptions (how to call this function), and it seems a bit easy to mishandle it, so I would strongly prefer a `Result<>`, despite a bit uglier to the users.

(Alternatively, can we clone the guard `MediaFileHandle` and consume it? :zany_face:)

Reply:
You wrote below:

> if the handle is missing after a call to persist, it's actually possible to crash the app via misuse of this API

but I think usually that wouldn't happen since panics are converted to exceptions at the FFI boundary and presumably there should be some catch-all boundaries for exceptions in most apps? I mean, even checked exceptions can cause the app to crash if they propagate up all the way to `main`.

Reply:
I still absolutely want this to not be a panic. Panic indicates an internal expectation being broken, and shouldn't happen because an API is misused.